### PR TITLE
Unify proj_create(), proj_create_from_user_input() and proj_create_from_proj_string() (fixes #1214)

### LINKED ISSUE #1214: Resolve differences with PJ objects created with proj_create versus PJ objects created with proj_create_from_user_input
PJ object created with `proj_create()` do not play well with the other functions that used to have the proj_obj namespace (i.e. objects created with `proj_create_from_user_input()`).

Also, PJ objects created with `proj_create_from_user_input()` do not play well with functions that were meant to be used with PJ objects created with `proj_create()`.

It would be good to merge this behavior when/if it makes sense.

Also, a for reference: https://github.com/OSGeo/proj.4/pull/1208#issuecomment-450754124

### PR DESCRIPTION
**Pull request by rouault**

### CODE REVIEW

File: docs/source/development/reference/functions.rst [hunk 1]
```diff
@@ -579,17 +579,16 @@ Various
 C API for ISO-19111 functionality
 +++++++++++++++++++++++++++++++++
 
-The PJ* objects returned by :c:func:`proj_create_from_user_input`,
-:c:func:`proj_create_from_wkt`, :c:func:`proj_create_from_proj_string`,
-:c:func:`proj_create_from_database` and other functions
+The PJ* objects returned by :c:func:`proj_create_from_wkt`,
+:c:func:`proj_create_from_database` and other functions in that sectino
```
Comment (kbevers) [inline docs/source/development/reference/functions.rst]: sectino -> section

File: docs/source/development/reference/functions.rst [hunk 2]
```diff
@@ -579,17 +579,16 @@ Various
...
 will have generally minimal interaction with the functions declared in the
 previous sections (calling those functions on those objects
 will either return an error or default/non-sensical values). The exception is
 for ISO19111 objects of type CoordinateOperation that can be exported as a
 valid PROJ pipeline. In this case,  objects will work for example with
 :c:func:`proj_trans_generic`.
-Conversely, objects returned by :c:func:`proj_create` and :c:func:`proj_create_argv` will
-return an error when used with functions of this section.
+Conversely, objects returned by :c:func:`proj_create` and :c:func:`proj_create_argv`,
+which are not of type CRS, will return an error when used with functions of this section.
```
Comment (kbevers) [inline docs/source/development/reference/functions.rst]: Would it make sense to add a note on how to test if a `PJ` is a CRS type?

File: src/proj_internal.h [hunk 3]
```diff
@@ -899,6 +899,11 @@ void   pj_freeup_plain (PJ *P);
 
 PJ* pj_init_ctx_with_allow_init_epsg( projCtx_t *ctx, int argc, char **argv, int allow_init_epsg );
 
+std::string PROJ_DLL pj_add_type_crs_if_needed(const std::string& str);
+
+PJ *proj_create_internal (PJ_CONTEXT *ctx, const char *definition);
```
Comment (kbevers) [inline src/proj_internal.h]: Per https://github.com/OSGeo/proj.4/issues/678, I suggest renaming to `pj_create_internal` and `pj_create_argv_internal`.

File: src/4D_api.cpp [hunk 4]
```diff
@@ -653,6 +656,38 @@ indicator, as in {"+proj=utm", "+zone=32"}, or leave it out, as in {"proj=utm",
     return P;
 }
 
+/*************************************************************************************/
+PJ *proj_create_argv_internal (PJ_CONTEXT *ctx, int argc, char **argv) {
```
Comment (kbevers) [inline src/4D_api.cpp]: Is the only difference from this one and the public counter-part that this one calls `proj_create_internal`?

**Review by kbevers — APPROVED** [review]
Seems good to me. Only a few minor comments. Apart from the ones inlined, it doesn't seem to that `+type=crs` is documented except within the code. Where would be a good place to put something on that? A general page describing the different interpretations of proj strings?

File: src/proj_internal.h (hunk 3 above)
Comment (rouault) [reply src/proj_internal.h]: why _argv ? Did you mean pj_create_internal ?

File: src/4D_api.cpp (hunk 4 above)
Comment (rouault) [reply src/4D_api.cpp]: yes

File: docs/source/development/reference/functions.rst (hunk 1 above)
Comment (rouault) [reply docs/source/development/reference/functions.rst]: Adressed

File: src/4D_api.cpp (hunk 4 above)
Comment (rouault) [reply src/4D_api.cpp]: Adressed by renaming proj_create_internal() to pj_create_internal() and proj_create_argv_internal() to pj_create_argv_internal()

File: src/proj_internal.h (hunk 3 above)
Comment (rouault) [reply src/proj_internal.h]: Adressed by renaming proj_create_internal() to pj_create_internal() and proj_create_argv_internal() to pj_create_argv_internal()

**Comment by rouault** [comment]
> Where would be a good place to put something on that? A general page describing the different interpretations of proj strings?

That's a good point. I'm not even sure there is a page explaining the basic syntax of a PROJ string. That would be the appropriate place. I'm not so good at writing doc unfortunately.

**Comment by kbevers** [comment]
> I'm not even sure there is a page explaining the basic syntax of a PROJ string. That would be the appropriate place.

We have something along the lines of that on https://proj4.org/usage/projections.html but it is definitely in need of a revision. I've been meaning to do that for a long time but I am struggling to find the time for it unfortunately. 

> I'm not so good at writing doc unfortunately.

You're plenty good, you just don't do it much :-)

**Comment by rouault** [comment]
Ah actually, the intro for proj strings is in https://proj4.org/usage/quickstart.html

**Pull request merged**